src/shared/bootspec.c

```diff
@@ -0,0 +1,335 @@
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Zbigniew Jędrzejewski-Szmek
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "bootspec.h"
+#include "conf-files.h"
+#include "def.h"
+#include "efivars.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "string-util.h"
+#include "strv.h"
+
+void boot_entry_free(BootEntry *entry) {
+        free(entry->filename);
+
+        free(entry->title);
+        free(entry->version);
+        free(entry->machine_id);
+        free(entry->architecture);
+        strv_free(entry->options);
+        free(entry->kernel);
+        free(entry->efi);
+        strv_free(entry->initrd);
+        free(entry->device_tree);
+}
+
+int boot_entry_load(const char *path, BootEntry *entry) {
+        _cleanup_fclose_ FILE *f = NULL;
+        unsigned line = 1;
+        _cleanup_(boot_entry_free) BootEntry tmp = {};
+        int r;
+
+        f = fopen(path, "r");
+        if (!f)
+                return log_error_errno(errno, "Failed to open \"%s\": %m", path);
+
+        tmp.filename = strdup(basename(path));
+        if (!tmp.filename)
+                return log_oom();
+
+        for (;;) {
+                _cleanup_free_ char *buf = NULL;
+                char *p;
+
+                r = read_line(f, LONG_LINE_MAX, &buf);
+                if (r == 0)
+                        break;
+                if (r == -ENOBUFS)
+                        return log_error_errno(r, "%s:%u: Line too long", path, line);
+                if (r < 0)
+                        return log_error_errno(r, "%s:%u: Error while reading: %m", path, line);
+
+                line++;
+
+                if (IN_SET(*strstrip(buf), '#', '\0'))
+                        continue;
+
+                p = strchr(buf, ' ');
```

Comment:
maybe use extract_first_word?

Reply:
I don't think so. The problem is that this must implement the exact same semantics as the efi code. So any support for quoting or escapes is counterproductive. When changed to use extract_first_word, the code is the same length, but there's the potential that all the bell and whistles of extract_first_word will cause some misparsing. I prefer to keep this as is.